ssl/record/rec_layer_s3.c

```diff
@@ -115,10 +115,22 @@ size_t ssl3_pending(const SSL *s)
     if (s->rlayer.rstate == SSL_ST_READ_BODY)
         return 0;
 
+    /* Take into account DTLS buffered app data */
+    if (SSL_IS_DTLS(s)) {
+        DTLS1_RECORD_DATA *rdata;
+        pitem *item, *iter;
+
+        iter = pqueue_iterator(s->rlayer.d->buffered_app_data.q);
+        while ((item = pqueue_next(&iter)) != NULL) {
+            rdata = item->data;
+            num += rdata->rrec.length;
+        }
+    }
+
     for (i = 0; i < RECORD_LAYER_get_numrpipes(&s->rlayer); i++) {
         if (SSL3_RECORD_get_type(&s->rlayer.rrec[i])
             != SSL3_RT_APPLICATION_DATA)
-            return 0;
+            return num;
```

Comment:
Wait, er, what?  Isn't this supposed to signal an error?  How does a value != 0 do that?

Reply:
No - this should not be an error. The return value represents the amount of application data that is buffered and available to read.

With DTLS there are 2 places application data might be buffered - either in the `buffered_app_data` queue - or in the `rrec` array. The `rrec` array could contain records that aren't app data. If that's the case then we ignore the data there. In that case we should return the amount of data we found in `buffered_app_data`.

Previously it was ok to just return 0 if we hit an `rrec` array entry that wasn't app data. All the `rrec` entries are guaranteed to be of the same type. So, if that type isn't app data, then we have 0 bytes of app data. Now that we're additionally checking `buffered_app_data` that is no longer the case.

Reply:
> SSL3_RT_APPLICATION_DATA records can't appear in rrec after records of other types, thus the loop can be terminated in this case.

Yes - exactly. Its not an error case